File: python/src/unclarity/scenario.py
from __future__ import annotations

from typing import Any


class ScenarioBuilder:
    """Fluent scenario builder; compiles to the same JSON wire form as the Node builder."""

    def __init__(self) -> None:
        self._steps: list[dict[str, Any]] = []

    def wait(self, ms: int) -> "ScenarioBuilder":
        self._steps.append({"type": "wait", "ms": ms})
        return self

    def move(self, selector: str) -> "ScenarioBuilder":
        self._steps.append({"type": "move", "selector": selector})
        return self

    def click(self, selector: str) -> "ScenarioBuilder":
        self._steps.append({"type": "click", "selector": selector})
        return self

    def scroll_to(self, y: int) -> "ScenarioBuilder":
        self._steps.append({"type": "scrollTo", "y": y})
        return self

    def type(self, selector: str, text: str) -> "ScenarioBuilder":
        self._steps.append({"type": "type", "selector": selector, "text": text})
        return self

    def build(self) -> dict[str, Any]:
        return {"steps": list(self._steps)}

    def to_json(self) -> dict[str, Any]:
        return self.build()


def scenario() -> ScenarioBuilder:
    return ScenarioBuilder()

File: unclarity-packages/core/src/scenario.ts
// Declarative, serializable scenario (the wire form shared with the Python client) + a fluent builder.
export type Step =
  | { type: "wait"; ms: number }
  | { type: "move"; selector: string }
  | { type: "click"; selector: string }
  | { type: "scrollTo"; y: number }
  | { type: "type"; selector: string; text: string };

export interface Scenario {
  steps: Step[];
}

export class ScenarioBuilder {
  private readonly steps: Step[] = [];
  wait(ms: number): this {
    this.steps.push({ type: "wait", ms });
    return this;
  }
  move(selector: string): this {
    this.steps.push({ type: "move", selector });
    return this;
  }
  click(selector: string): this {
    this.steps.push({ type: "click", selector });
    return this;
  }
  scrollTo(y: number): this {
    this.steps.push({ type: "scrollTo", y });
    return this;
  }
  type(selector: string, text: string): this {
    this.steps.push({ type: "type", selector, text });
    return this;
  }
  build(): Scenario {
    return { steps: [...this.steps] };
  }
  toJSON(): Scenario {
    return this.build();
  }
}

export function scenario(): ScenarioBuilder {
  return new ScenarioBuilder();
}
